windows/lcow_runtime.go

```diff
@@ -0,0 +1,400 @@
+	// Store the LayerFolder in the db so we can clean it if we die
+	if err = r.db.Update(func(tx *bolt.Tx) error {
+		s := newLayerFolderStore(tx)
+		return s.Create(nsid, conf.LayerFolderPath)
+	}); err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			if dbErr := r.db.Update(func(tx *bolt.Tx) error {
+				s := newLayerFolderStore(tx)
+				return s.Delete(nsid)
+			}); dbErr != nil {
+				log.G(ctx).WithField("id", id).
+					Error("failed to remove key from metadata")
+			}
+		}
+	}()
+
+	ctr, err := hcsshim.CreateContainer(nsid, conf)
+	if err != nil {
+		time.Sleep(15 * time.Second)
+		return nil, errors.Wrapf(err, "hcsshim failed to create task")
+	}
+	defer func() {
+		if err != nil {
+			ctr.Terminate()
+			ctr.Wait()
+			ctr.Close()
+		}
+	}()
+
+	if err = ctr.Start(); err != nil {
+		return nil, errors.Wrap(err, "hcsshim failed to spawn task")
+	}
+
+	t := &task{
+		id:                id,
+		namespace:         namespace,
+		pid:               pid,
+		io:                pset,
+		status:            runtime.CreatedStatus,
+		spec:              spec,
+		processes:         make(map[string]*process),
+		publisher:         r.publisher,
+		rwLayer:           conf.LayerFolderPath,
+		pidPool:           r.pidPool,
+		isWindows:         false,
+		hyperV:            true,
+		hcsContainer:      ctr,
+		terminateDuration: createOpts.TerminateDuration,
+	}
+	r.tasks.Add(ctx, t)
+
+	var rootfs []*containerdtypes.Mount
+	for _, l := range append([]string{t.rwLayer}, spec.Windows.LayerFolders...) {
+		rootfs = append(rootfs, &containerdtypes.Mount{
+			Type:   "lcow-layer",
+			Source: l,
+		})
+	}
+
+	r.publisher.Publish(ctx,
+		runtime.TaskCreateEventTopic,
+		&eventsapi.TaskCreate{
+			ContainerID: id,
+			IO: &eventsapi.TaskIO{
+				Stdin:    io.Stdin,
+				Stdout:   io.Stdout,
+				Stderr:   io.Stderr,
+				Terminal: io.Terminal,
+			},
+			Pid:    t.pid,
+			Rootfs: rootfs,
+			// TODO: what should be in Bundle for windows?
+		})
+
+	return t, nil
+}
+
+func (r *LCOWRuntime) cleanup(ctx context.Context) {
+	cp, err := hcsshim.GetContainers(hcsshim.ComputeSystemQuery{
+		Types:  []string{"Container"},
+		Owners: []string{hcsshimOwner},
+	})
+	if err != nil {
+		log.G(ctx).Warn("failed to retrieve running containers")
+		return
+	}
+
+	for _, p := range cp {
+		container, err := hcsshim.OpenContainer(p.ID)
+		if err != nil {
+			log.G(ctx).Warnf("failed open container %s", p.ID)
+			continue
+		}
+
+		err = container.Terminate()
+		if err == nil || hcsshim.IsPending(err) || hcsshim.IsAlreadyStopped(err) {
+			container.Wait()
+		}
+		container.Close()
+
+		// TODO: remove this once we have a LCOW snapshotter
+		var layerFolderPath string
+		if err := r.db.View(func(tx *bolt.Tx) error {
+			s := newLayerFolderStore(tx)
+			l, e := s.Get(p.ID)
+			if err == nil {
+				layerFolderPath = l
+			}
+			return e
+		}); err == nil && layerFolderPath != "" {
+			removeLinuxLayer(ctx, layerFolderPath)
```

Comment:
Hmmm, thinking more about this, I think each runtime should have a separate owner to prevent them stepping on each other's toes during the initial cleanup. I'll combine it too.